### DISCUSSION TITLE
Allow parameterless (except `this` parameter) extension methods to behave as properties to participate in pattern matching

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#3998] [Language Ideas]
For example, consider the code:
```csharp
// To check whether the value is power of two.
// This is an extension method of type 'System.Int32'.
public static bool IsPowerOfTwo(this int @this) => @this != 0 && (@this & (@this - 1)) == 0;
```

In this way, it's unnecessary to create a new syntax to describe so-called "extension properties". We use the parameterless extension methods for two purposes:
1. To describe them as normal methods that is just like the members belonging to certain types.
2. To use them on pattern matching (duck types) like
```csharp
// Here 'variable' may be other types. (How about 'System.Object'?)
if (variable is int { IsPowerOfTwo: true } v)
{
    // Code using 'v'.
}
```

Thanks!

**Comment by alrz** [comment]
See https://github.com/dotnet/csharplang/issues/1047. it will enable this one way or another, though the syntax or semantics is not finalized yet.

**Reply by kyoyama-kazusa** [reply]
Thank you for your reply. I've seen that issue. `operator is` is a extremely powerful syntax! But... I think that `operator is` is a little bit... weird? `operator is` can only make a specified value as the right (`v is SomeMethodAndReturnsAResult()`) rather than a pattern `{ Prop: ConstantValue }` or a variable defination (`v is int i`) (Or maybe I miss the point).

Of course, we can now use `is` pattern to check whether the instance is the specified type and the specified value. However, if we create a user-defined `operator is`, will there be any effects to break the basic checking?
